OAK-10590 - If includedPaths and excludedPaths are specified as a String instead of array of String, interpret them as a one-element array of Strings

nfsantos opened the pull request:
The includedPaths property of an index definition should be an array of strings. But a common mistake made by users is to define it as a String when it has a single element. That is, instead of:
```
        "includedPaths": [ "/a/b"] , 
```
it is defined as:
```
        "includedPaths": "/a/b", 
```

If `includedPaths` is defined as a String, the indexing job would ignore its value and instead default to use the root as the `includedPaths`, which results in downloading the full node store and creating an FFS containing everything (except hidden paths). This will slow down significantly the indexing job, as it will negate any benefits from using regex filtering. And even if regex filtering is not enabled or cannot be used, using `/` as `includedPaths` will also result in the FFS containing more nodes than it should, which will once again slow down the indexing job.

The same is true for `excludedPaths`, but in this case, the default value is an empty list of `excludedPaths`, so it will ignore the value of this property and will not exclude anything. This may result in parts of the node store being indexed that should not be indexed.

The handling of `includedPaths` and `excludedPaths` is also inconsistent with the handling of `queryPaths`, which is being interpreted as a one-element array if it is defined as a String.

This PR makes the logic that reads the `includedPaths` and `excludedPaths` properties more lenient, by treating Strings as one-element arrays and issuing a warning with a suggested fix.

Additionally, the PR makes some minor cleanups in the files that had to be modified (remove usages of Guava and fix some compilation warnings).

nfsantos committed:
WIP

nfsantos committed:
If an index definition has includedPaths or excludedPaths defined as a String instead of array of String, log a warning and treat them as a one-element array instead of assuming that it they are not defined.

nfsantos committed:
Fix wrong commit.

thomasmueller commented on oak-store-spi/src/main/java/org/apache/jackrabbit/oak/spi/filter/PathFilter.java:
```diff
@@ -164,26 +167,36 @@ public String toString() {
             return property.getValue(Type.STRINGS);
+        } else if (property != null && property.getType() == Type.STRING) {
+            String value = property.getValue(Type.STRING);
+            LOG.warn("Property \"{}\"=\"{}\" has type String but it should be array of String. Proceeding by treating it as a " +
```
I would probably not log a warning. Instead, I would document that this is also supported.

Reason: we already have many many entries that are defined as string (mostly "/dummy") instead of array.

thomasmueller replied:
I would make it public and add a unit test case for it.

nfsantos committed:
Remove setting excludePaths which is not relevant for this test and was being done incorrectly, by setting the property as a string instead of array of strings.

nfsantos committed:
Update all parts of the code that read included/excluded paths to accept a string value as being a one-element array (NodeCounterMBeanEstimator).
Other refactoring to clean up the code.

nfsantos committed:
Add tests and reduce code duplication.

nfsantos committed:
Add test to check that queryPaths is accepted if it is defined as a STRING instead of STRINGS.
Simplified the logic of extracting a list of strings from a property that can be STRING or STRINGS by taking advantage of the default value of PropertyState.getValue(STRINGS), which handles the case of the property being a STRING.
